### PR TITLE
Document rustfmt on nightly-rustc

### PR DESCRIPTION
**Pull request by jyn514**
- Refactor the doc step for Rustdoc into a macro
- Call the macro for both rustdoc and rustfmt
- Add a `recursion_limit` macro to avoid overflow errors

This does not currently pass --document-private-items for rustfmt due to https://github.com/rust-lang/cargo/pull/8422#issuecomment-871082935.

r? @Mark-Simulacrum cc @calebcartwright

### CODE REVIEW

**Comment by jyn514** [comment]
See https://rust-lang.zulipchat.com/#narrow/stream/233931-t-compiler.2Fmajor-changes/topic/rustdoc.20is.20using.20rustc_ast_pretty.2C.20would.20.E2.80.A6.20compiler-team.23403 for the motivation - rustdoc is considering using rustfmt as a library.

File: src/bootstrap/doc.rs [hunk 1]
```diff
@@ -593,84 +593,91 @@ impl Step for Rustc {
...
+            }
 
-        // Build rustc docs so that we generate relative links.
-        builder.ensure(Rustc { stage, target });
+            fn make_run(run: RunConfig<'_>) {
+                run.builder.ensure($tool { stage: run.builder.top_stage, target: run.target });
+            }
 
-        // Build rustdoc.
-        builder.ensure(tool::Rustdoc { compiler });
```
Comment (jyn514) [inline src/bootstrap/doc.rs:641]: This wasn't actually necessary - rustdoc is built in the step before when `Rustc` is documented.

**Comment by Mark-Simulacrum** [comment]
Seems reasonable to me -- r=me with CI fixed (looks like rustfmt failure).

**Pull request merged**